### PR TITLE
Add servlet for handling user authentication

### PR DESCRIPTION
**Pull request by CarlosM10011**
This pull request provides a servlet for JS to use when logging in or out.

partially addresses issue #12.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -30,21 +34,49 @@
...
+    }
     String output = gson.toJson(this.comments);
-    response.setContentType(this.CONTENT_TYPE);
+    response.setContentType(this.GET_CONTENT_TYPE);
     response.getWriter().println(output);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final String name = request.getParameter(this.POST_NAME_PARAMETER);
```
Comment (hzimz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Just POST_NAME_PARAMETER

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  private final String CONTENT_TYPE = "application/json;";
+  private final String DEFAULT_REDIRECT_URL = "/";
+  private final Gson gson = new Gson();
+  private final String REDIRECT_URL_PARAMETER_NAME = "redirect";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final AuthState authState = AuthStateFactory.getState();
+    final String redirectUrl = request.getParameter(this.REDIRECT_URL_PARAMETER_NAME);
+    if (redirectUrl != null) {
```
Comment (hzimz) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Can shorten this to smth ```authState.setRedirectUrl(redirectUrl != null ? redirectUrl : DEFAULT_REDIRECT_URL)```

**Review by hzimz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles authentication. */
+@WebServlet("/auth")
+public class AuthServlet extends HttpServlet {
+
+  private final String CONTENT_TYPE = "application/json;";
+  private final String DEFAULT_REDIRECT_URL = "/";
```
Comment (tinagrasso) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Are these duplicated in other files? You should create a file to hold constants like these if so.

**Review by tinagrasso — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -30,21 +34,49 @@
...
+    PreparedQuery results = this.datastore.prepare(this.query);
+    for (Entity comment : results.asIterable()) {
+      final String name = (String) comment.getProperty(PROPERTY_COMMENT_NAME);
+      final Date created = new Date((long) comment.getProperty(PROPERTY_COMMENT_CREATED_NAME));
+      final String message = (String) comment.getProperty(PROPERTY_COMMENT_MESSAGE_NAME);
+      this.comments.add(new Comment(name, created, message));
+    }
     String output = gson.toJson(this.comments);
-    response.setContentType(this.CONTENT_TYPE);
+    response.setContentType(this.GET_CONTENT_TYPE);
```
Comment (hzimz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Same as before, just use GET_CONTENT_TYPE

**Review by hzimz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java (hunk 3 above)
Comment (CarlosM10011) [reply portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Yes. I moved them into a new file.

Comment (tinagrasso) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Remove and use the constant you just created.

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that handles authentication. */
+@WebServlet("/auth")
+public class AuthServlet extends HttpServlet {
+
+  private final String CONTENT_TYPE = "application/json;";
```
Comment (tinagrasso) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Group instance variables in some logical way. Read the section in the style guide: https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents

Also make string constants static.

**Review by tinagrasso — CHANGES_REQUESTED** [review]

**Review by hzimz — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java (hunk 3 above)
Comment (CarlosM10011) [reply portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: Sorry, I overlooked this file. I moved all of them into the constants file.

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java (hunk 5 above)
Comment (CarlosM10011) [reply portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java]: I rearranged the constants and added comment headers to the constants file.

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+public class AuthServlet extends HttpServlet {
+
+  private final Gson gson = new Gson();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final AuthState authState = AuthStateFactory.getState();
+    final String redirectUrl = request.getParameter(Constants.REDIRECT_AUTH_URL_PARAMETER_NAME);
+    authState.setRedirectUrl(
+        redirectUrl != null ? redirectUrl : Constants.DEFAULT_AUTH_REDIRECT_URL);
```
Comment (tinagrasso) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java:37]: Consider wrapping with `Optional` to protect against null pointer exceptions.

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html

**Review by tinagrasso — APPROVED** [review]
